### PR TITLE
[combobox] Fix issue with Combobox context not being memoized correctly

### PR DESCRIPTION
**Pull request by Andarist**
This pull request:
- [x] Fixes a bug in an existing package

When trying to reproduce a problem described in https://github.com/reach/reach-ui/issues/199 I've noticed that it currently cannot be reproduced because:
- [`transition`](https://github.com/reach/reach-ui/blob/f22cdf931d52384205575f3060933241056e4d1e/packages/combobox/src/index.js#L813) is an inline function, so it has a fresh reference for each render
- but it's used as input for `useMemo` [here](https://github.com/reach/reach-ui/blob/f22cdf931d52384205575f3060933241056e4d1e/packages/combobox/src/index.js#L282), which makes `useMemo` yield a new reference each render and thus constantly updating the context value

I've fixed it by creating a "proxy" callback to this inline one. This way that "proxy" can have a stable reference across renders but still call "latest" version of the inline function thanks to using a ref under the hood.

I believe this solution is better than using `useCallback` directly because I think that `transition`'s reference shouldn't ever update, it should be considered stable - similarly how we can assume `setState` & `dispatch` to be stable.

### CODE REVIEW

File: packages/utils/src/index.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
-import { useRef, useMemo, useEffect } from "react";
+import { useRef, useMemo, useEffect, useCallback } from "react";
+import useLatest from "use-latest";
```
Comment (Andarist) [inline packages/utils/src/index.js]: If you want - I can inline this dependency. It's really simple but my personal opinion is that it's good to share common code (even if it's short like this one), but I know many people prefer inlining for cases like this. Disclaimer though - it's written by me 😉 I actually also probably will extract `useStableCallback` to a separate library as well, just didn't have time to do this right now.

**Comment by chancestrickland** [comment]
I'm still unsure what bug this is actually addressing. So you're saying that, if we merge this PR, then we'll have a bug in combobox per #199? But we don't currently have any bugs in combobox described in #199, is that correct? What problem does this PR ultimately solve for us today?

**Comment by Andarist** [comment]
> What problem does this PR ultimately solve for us today?

I assume this useMemo:
https://github.com/reach/reach-ui/blob/f22cdf931d52384205575f3060933241056e4d1e/packages/combobox/src/index.js#L265
has been introduced for a reason. Right now you can either fix its dependencies (by merging this PR or some alternative way of fixing this `useMemo` dependencies) or remove this `useMemo` entirely.

> So you're saying that, if we merge this PR, then we'll have a bug in combobox per #199? But we don't currently have any bugs in combobox described in #199, is that correct?

I've found out about this issue here when I've tried to repro #199 in a very similar way as I've been reproducing it in the past - when `useMemo` was memoized correctly. When you force whole combobox-related stuff to rerender on each change then the reported issue goes away as it was specifically about memoized components in the middle breaking things. Question is - do you want to rerender everything in unrelated changes?

**Comment by chancestrickland** [comment]
@Andarist Understood, thanks for your explanation and the PR!

I think for now, to keep it simple, we are just going to remove `useMemo` altogether. The biggest reason is because we will be rolling out a rewrite of combobox in the coming weeks, and adding complexity + a new dependency here is less desirable ATM. We can circle back to this approach if we need it for the optimization once the next version is released!

**Comment by Andarist** [comment]
@chancestrickland cool, good to see that useMemo removed for now 👍 When you get to refactoring this component, I would appreciate giving me a ping - would love to see how you are going to do it 😃

**Pull request closed without merging**